mkdocs_simple_plugin/plugin.md

```diff
@@ -14,25 +14,114 @@ _Python 3.x, 3.5, 3.6, 3.7, 3.8, 3.9 supported._
       include_folders: ["*"]
-      # Optional setting to ignore specific folders
-      ignore_folders: [""]
-      # Optional setting to specify if hidden folders should be ignored
+      # In addition, directories whose name matches a glob pattern in this
+      # list will NOT be searched for documentation.
+      ignore_folders: ["site"]
+      # Hidden directories will not be searched if this is true:
       ignore_hidden: True
-      # Optional setting to specify other extensions besides md files to be copied
+      # If the following is true, the contents of the docs directory (if any)
+      # will be merged at the same level as all other documentation, i.e., the
+      # docs directory prefix of these files will be elided. Otherwise, the docs
+      # directory will be searched like all other directories, i.e., the docs
+      # directory prefix will be kept:
+      merge_docs_dir: True
+      # Any file in the searched directories whose name contains a string in
+      # this list will simply be copied to the generated documentation:
       include_extensions: [".tif", ".tiff", ".gif", ".jpeg", ".jpg", ".jif", ".jfif",
             ".jp2", ".jpx", ".j2k", ".j2c", ".fpx", ".pcd", ".png", ".pdf", "CNAME"]
-      # Optional setting to merge the docs directory with other documentation
-      merge_docs_dir: True
+      # The dicts in the following list specify which files will be scanned for
+      # embedded documentation and how. The first entry in the list gives all
+      # of the keys allowed in an entry with their meanings and default values.
+      # The remaining entries in the list actually constitute the default value
+      # of the "semiliterate" parameter:
+      semiliterate:
+          # Any file in the searched directories whose name contains this
+          # required regular expression parameter will be scanned:
+        - pattern: None
+          # However, if the file has already matched one of the
+          # "include_extensions" or an earlier semiliterate pattern, it will not
+          # be scanned again unless this parameter is true:
+          rescan: false
+          # Extracted documentation will be copied to a file whose name is
+          # generated by replacing the last matching group in the "pattern"
+          # parameter with this string:
+          to_extension: "md"
+          # If the following parameter is specified, lines from the scanned file
+          # will be ignored until a line containing this regular expression is
+          # encountered. As with all of the remaining parameters, the last
+          # matching group in the expression, if any, will be written to the
+          # extracted file. Subsequent lines, or lines from the beginning of the
+          # file if "start" is not specified, possibly with the transformations
+          # specified by the "replace" parameter below, will be written to
+          # the extracted file.
+          start: None
```

Comment:
I assume you mean that the 'begin' marker would serve as both start and resume, and that 'end' would serve as pause? I could switch to that, but this is a bit more flexible, in that: (A) by not specifying start but only pause and resume, you can indicate that extraction should start from the beginning of the file; and (B) you can have a marker that completely terminates (or even pre-empts) extraction from the file.

Since this was to be as language-agnostic as possible, I wanted as full flexibility in extracting snippets of code or documentation from arbitrary source files as possible, while keeping to fairly intuitive meanings, which I think start, pause, resume, end do. In my projects so far I have used start/pause, start/end, and pause/resume pairs productively.

Of course the PR could be adjusted. The easiest change would be to eliminate "resume," and merge its meaning with "start." As far as I can tell from my experience, the only real loss would be not having a way to start extraction from the first line of a file. I've never actually used a value of "resume" different from that for "start," which is why I put in the default for resume. (So I could also adjust the PR to have "start," "pause," and "end," plus a flag for whether the initial state is "extracting" or "not extracting" -- that's only 3 regexp parameters then and one flag, which is arguably simpler than 4 regexps. But I'd prefer to have both "pause" and "end" if you don't mind, their semantics are quite different, and end is useful because then you just don't have to worry about what the contents of the rest of the file are, you won't trigger unwanted extraction.)

Reply:
Yes, I mean that it seems like "resume/pause" is the more useful parameters.

I'm not sure I understand the use case of wanting to grab from the beginning of a file and terminating halfway through?  What case are you trying to solve by 'end' that is not covered by 'pause'?

Reply:
I was trying to provide user flexibility in the face of unknown languages/codebase arrangements that someone might be extracting documentation from.

For example, one could adopt a convention that there's Markdown in the opening comment block of (some kind of) scripts with `#`-marked line comments. Then one could write:
```
- pattern: '(.script)'      # whatever your script extension is
  replace: ['^#(.*\s*)$'] # strip leading `#`
  end: '^[^#]'                # terminate processing at end of beginning comment block
```
Then you can use comments normally in the rest of your script without worrying about accidentally creating more markdown. In general, since this proposed extraction facility for `simple` is operating via ad-hoc pattern matching, I see value in being able to terminate processing of a file altogether, so that below the end marker, totally free usage of whatever the file's primary language syntax is possible without worry of triggering additional markdown extraction.

Also, I was trying to create common parameters for markdown extraction and for the `{! ... !}` inclusion facility, which I so far have mainly used to extract source code examples into the documentation. When including, I generally have in mind "end" semantics rather than "pause" -- I am trying to extract one specific code snippet, and when I hit its end marker, I want extraction to stop. I don't want to inadvertently get more code because the start marker I chose happens to occur again later in the source file.

One other small example: in the semiliterate package, I defined 'end' for markdown files to be '<!-- /md -->'. That way, I could have a block at the end of README files explaining where to get further information, which would show in the repository view, but which is unnecessary and would be dropped in the documentation site. This is a fairly weak example, as 'pause' would work in practice there too, but the semantics I had in mind was "Hey, mkdocs, there's nothing else of interest to you in this file." The end parameter captures that, whereas pause at least suggests that there may be more content to come.

Reply:
Ah, I see.  What you're describing is more akin to "front matter", whereas what I was expecting was more like a "code block".

I think it's fair to assume that people might comment their code in one of two styles
(1) The "front matter" style, where documentation is written at the top of the file.
(2) With "documentation blocks", where documentation is written throughout.

TL;DR I went back and forth on this since it seems like it should be reasonable to assume that the user can come up with a unique end string, but I'm ok with leaving a terminator.

From a philosophical point of view:

The main way in which people comment their code is via language standards.  Each one is unique and has a different syntax.  These "standard" versions of commenting code are well covered in other plugins (sphynx, mkdocstrings).  However, it would be more convenient to be able to write markdown documentation directly into source.  This is the purpose of this feature, it adds the ability to search through files for markdown snippets.  Therefore, it is reasonable to assume that users are going to be modifying the way they comment their code in order to use this feature.

Since users are already doing something against language standards to use this feature, I think it's reasonable to require a unique start/end tag.  I'm having a hard time coming up with examples in which makes finding a unique termination string difficult.  

The "front matter" case would mean that a user would only need to worry about the first instance of a string before terminating documentation, letting the first non-comment line signify the end of front matter documentation seems like a usable feature, even though it wouldn't be "unique".

Reply:
Thanks for putting so much thought into this! Here is where I _think_ you landed: extraction parameters "begin," "pause," and "end," with files always starting in "non-extracting" mode, both of what I called "start" and "resume" merged into "begin", and "pause" and "end" keeping the semantics I proposed. Then "documentation blocks" can be done by a combination of "begin" and "pause", and "front matter" can be done by a combination of setting "begin" to `^(.*\s*)$` and using "end".

I think that actually covers all of my current use cases, so if I have followed you correctly I am perfectly happy with this as a landing place. I will get started with all of the other items, and please confirm my understanding. Also, please weigh in if you have thoughts on the following point given where we've landed:

* Should "end" occurring before the first occurrence of "begin" pre-empt processing altogether? Right now it does, but in this setup I could go either way -- rather than having the occurrence of "end" pre-empt "begin," it could simply be the non-presence of "begin" that leads to nothing extracted from a file. That might make it easier when grabbing a code snippet as an example in documentation, to choose a known well-identified starting point (like a function name) and then close it at a possibly generic spot (like a bracket in the leftmost column) without being stymied because that end marker also occurs earlier in the file.

Thanks for considering.

Reply:
> Thanks for putting so much thought into this! Here is where I think you landed: extraction parameters "begin," "pause," and "end," with files always starting in "non-extracting" mode, both of what I called "start" and "resume" merged into "begin", and "pause" and "end" keeping the semantics I proposed. Then "documentation blocks" can be done by a combination of "begin" and "pause", and "front matter" can be done by a combination of setting "begin" to ^(.*\s*)$ and using "end".

Yes, that sounds correct to me

>Should "end" occurring before the first occurrence of "begin" pre-empt processing altogether? Right now it does, but in this setup I could go either way -- rather than having the occurrence of "end" pre-empt "begin," it could simply be the non-presence of "begin" that leads to nothing extracted from a file. That might make it easier when grabbing a code snippet as an example in documentation, to choose a known well-identified starting point (like a function name) and then close it at a possibly generic spot (like a bracket in the leftmost column) without being stymied because that end marker also occurs earlier in the file.

I don't think it matters much.  If end == terminate extraction then nothing will get extracted, if begin doesn't occur then also nothing will get extracted, right?  I think both things can be true.  

I think you can grab your code snippet with "begin" and "pause", as there would be just one pair of those in the file (and your pause marker should be the first occurrence of that match after begin).